src/autoPagination.ts

```diff
@@ -92,39 +78,97 @@ export function makeAutoPaginationMethods(
       );
     }
 
-    if (i < pageResult.data.length) {
-      const idx = reverseIteration ? pageResult.data.length - 1 - i : i;
+    const reverseIteration = isReverseIteration(this.requestArgs);
+    if (this.i < pageResult.data.length) {
+      const idx = reverseIteration
+        ? pageResult.data.length - 1 - this.i
+        : this.i;
       const value = pageResult.data[idx];
-      i += 1;
+      this.i += 1;
 
       return {value, done: false};
     } else if (pageResult.has_more) {
       // Reset counter, request next page, and recurse.
-      i = 0;
-      pagePromise = getNextPagePromise(pageResult);
-      return pagePromise.then(iterate);
+      this.i = 0;
+      this.pagePromise = this.getNextPage(pageResult);
+      return this.pagePromise.then((pageResult) => this.iterate(pageResult));
```

Comment:
can we use `await` instead of callback chaining?

Reply:
Ultimately yes, but if I made `iterate` an async function then the behavior of `throw` on line 76 would change. It would probably change for the better (it should be rejecting a promise instead of throwing outside the promise chain) but I wanted to keep this PR a true refactor.

Reply:
Done!